flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java

```diff
@@ -236,8 +273,16 @@ private void flush() {
                                 "Mark in-flight request as completed and requeue %d request entries",
                                 failedRequestEntries.size());
 
+        Consumer<Exception> fatalExceptionCons =
+                exception ->
+                        mailboxExecutor.execute(
+                                () -> {
+                                    throw exception;
+                                },
+                                "A fatal exception occurred in the sink that cannot be recovered from or should not be retried.");
+
         inFlightRequestsCount++;
-        submitRequestEntries(batch, requestResult);
+        submitRequestEntries(batch, requestResult, fatalExceptionCons);
```

Comment:
I think we should isolate this change on a separate commit because it changes framework code and is not only specific to kinesis.

Reply:
Sound good to me !